enhancements/node-tuning/containerize-tuned.md

```diff
@@ -0,0 +1,255 @@
+---
+title: containerize-tuned
+authors:
+  - "@yanirq"
+  - "@jmencak"  
+reviewers:
+  - "@MarSik"
+  - "@ffromani"
+  - "@Tal-or"
+  - "@browsell"
+  - "@bwensley"
+  - "@swatisehgal"
+  - "@mrunalp"  
+approvers:
+  - "@MarSik"
+  - "@browsell"
+  - "@mrunalp"
+api-approvers:
+  - "None"
+creation-date: 2024-03-11
+last-updated: 2024-03-14
+tracking-link:
+  - https://issues.redhat.com/browse/PSAP-1236
+---
+
+# Containerize Tuned
+
+## Summary
+
+The proposed enhancement is to restructure cluster-node-tuning-operator (NTO) to run TuneD from a container (e.g. using podman from a systemd service) instead of running in daemon mode so TuneD will initially set defaults and hand off ownership to other services that apply tunings.
+
+## Motivation
+
+With the current implementation of [cluster-node-tuning-operator](https://github.com/openshift/cluster-node-tuning-operator) (NTO) having TuneD running in a daemonset there is an overarching issue where we have multiple services manipulating tuning data - tuned, crio and some system services for irqs where there should only be one.
+
+This is manifested already several issues: 
+
+* Order of pods start up ([OCPBUGS-26401](https://issues.redhat.com/browse/OCPBUGS-26401)) : In case of node reboot, all pods restart again in random order. Since there's no control on pod restart order, it is possible that TuneD pod will start after the workloads. This means the workload pods start with partial tuning which can affect performance or even cause the workload to crash. One of the reasons being partial tuning and/or dynamic tuning done by crio hooks is undone.
+
+* Tuned restarts / reapplication ([OCPBUGS-26400](https://issues.redhat.com/browse/OCPBUGS-26400) , [OCPBUGS-27778](https://issues.redhat.com/browse/OCPBUGS-27778) , [OCPBUGS-22474](https://issues.redhat.com/browse/OCPBUGS-22474)) : If TuneD restarts at runtime for any reason - tuning is broken.
+
+* Tuning changes impact running applications ([OCPBUGS-28647](https://issues.redhat.com/browse/OCPBUGS-28647))
+
+* Historical issues:
+  * Tuned overwriting IRQBALANCE_BANNED_CPUS ([OCPBUGSM-46474](https://issues.redhat.com/browse/OCPBUGSM-46474))
+  * Tuned affining containers to house keeping cpus ([OCPBUGSM-31773](https://issues.redhat.com/browse/OCPBUGSM-31773))
+
+
+### User Stories
+
+No new user stories added in this EP.
+
+### Goals
+
+* Tuning integrity is kept and not corrupted when node restarts occur
+* Tuning integrity is kept and not corrupted when pod restarts occur
+* Tuning will be done as “one shot” to the majority of use cases
```

Comment:
Will this EP include a change to stop tuned profiles from being re-applied when the profile changes? If so, then I think that would enforce the "one shot" for all use cases - right? If you are not including that change, then what is enforcing the "one shot" aspect of this?

Reply:
So we are looking into that - e.g. solve at the bug level with [OCPBUGS-26401](https://issues.redhat.com/browse/OCPBUGS-26401) and [OCPBUGS-28647](https://issues.redhat.com/browse/OCPBUGS-28647) . 

If the implementation intersects with this enhancement then we might indeed be enforcing "one shot" here (excluding possible [TuneD runtime reloads](https://github.com/openshift/enhancements/pull/1597/files#r1529709349)).

@jmencak  @ffromani  any additional thoughts here ?

Reply:
A suggested solution by @jmencak  to be done before this EP: 
Remove the Tuned "rendered" resource and augment the TunedProfiles with a list of Tuned profiles (Spec.Profile) Profile's Spec.Config.TunedProfile was calculated out of.  This is to unlink profile changes from tuned changes when they are not necessarily needed in order to reduce tuned reloads.

Reply:
This is a key component to nail down.

1. Are we only allowing one-shot period ? Any new profile changes require a reboot ala a mc ?
2. Are we allowing one-shot for some tunings/plugins but runtime changes for others ? If so how is that achieved with the service approach ?